src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -95,8 +97,289 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
+                        default:
+                            Debug.Fail("Array.Copy from I4 to another type hit unsupported widening conversion"); break;
+                    }
+                    break;
+
+                case CorElementType.ELEMENT_TYPE_U8:
+                    switch (destElType)
+                    {
+                        case CorElementType.ELEMENT_TYPE_R4:
+                            GenericScalarWiden<ulong, float>(sourceArray, sourceIndex, destinationArray, destinationIndex, length); break;
+                        case CorElementType.ELEMENT_TYPE_R8:
+                            GenericScalarWiden<ulong, double>(sourceArray, sourceIndex, destinationArray, destinationIndex, length); break;
+                        default:
+                            Debug.Fail("Array.Copy from U8 to another type hit unsupported widening conversion"); break;
+                    }
+                    break;
+
+                case CorElementType.ELEMENT_TYPE_I8:
+                    switch (destElType)
+                    {
+                        case CorElementType.ELEMENT_TYPE_R4:
+                            GenericScalarWiden<long, float>(sourceArray, sourceIndex, destinationArray, destinationIndex, length); break;
+                        case CorElementType.ELEMENT_TYPE_R8:
+                            GenericScalarWiden<long, double>(sourceArray, sourceIndex, destinationArray, destinationIndex, length); break;
+                        default:
+                            Debug.Fail("Array.Copy from I8 to another type hit unsupported widening conversion"); break;
+                    }
+                    break;
+
+                case CorElementType.ELEMENT_TYPE_R4:
+                    Debug.Assert(destElType == CorElementType.ELEMENT_TYPE_R8);
+                    GenericScalarWiden<float, double>(sourceArray, sourceIndex, destinationArray, destinationIndex, length); break;
+
+                default:
+                    Debug.Fail("Fell through outer switch in PrimitiveWiden!  Unknown primitive type for source array!"); break;
+            }
+
+            static void GenericScalarWiden<TFrom, TTo>(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length)
+                where TFrom : unmanaged, INumber<TFrom>
+                where TTo : unmanaged, INumber<TTo>
+            {
+                ref TFrom src = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<TFrom[]>(sourceArray)), sourceIndex);
+                ref TTo dst = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<TTo[]>(destinationArray)), destinationIndex);
+
+                for (int i = 0; i < length; i++)
+                {
+                    Unsafe.Add(ref dst, i) = TTo.CreateTruncating(Unsafe.Add(ref src, i));
+                }
+            }
+        }
```

Comment:
This looks attractive to vectorize. However there are several concerns:

- The vectorized logic isn't quite the same for each src-dst combination. Vectorization for each case would cause significant code bloat.
- The logic feels better for TensorPrimitives, which can also define better strongly-typed API.
- Array.Copy isn't a common path like WidenAsciiToUtf16.

Thus, I choose to leverage generic math to create a general scalar implementation.

Reply:
This is trying to be faster than the existing unmanaged implementation, at the cost of introducing code bloat. I do not think this needs to be optimized. This path only exists for compatibility, it is very rarely used, and it is fine for it to be slow.

Reply:
The NativeAOT implementation of this path is very similar to unmanaged implementation. I would compare them and prefer code size here.